storage: TestRangeTransferLeaseExpirationBased failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/0f5a289d49105cd317a2713b71546d84c25547f6

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestRangeTransferLeaseExpirationBased PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=938214&tab=buildLog

```
I181002 08:09:56.478336 78670 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:33193
I181002 08:09:56.484607 78670 gossip/client.go:134  [n2] closing client to node 1 (127.0.0.1:33193): stopping outgoing client to node 1 (127.0.0.1:33193); already have incoming
W181002 08:09:56.534737 78481 storage/store.go:1651  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181002 08:09:56.719042 78386 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181002 08:09:57.149556 78387 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181002 08:09:57.393332 78481 storage/store.go:1651  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181002 08:09:57.443304 78388 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
I181002 08:09:57.479550 78673 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:33193
I181002 08:09:57.484893 78673 gossip/client.go:134  [n2] closing client to node 1 (127.0.0.1:33193): stopping outgoing client to node 1 (127.0.0.1:33193); already have incoming
W181002 08:09:57.511163 78482 storage/store.go:1651  [s2,r1/2:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181002 08:09:57.605328 78386 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181002 08:09:58.045256 78387 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181002 08:09:58.362360 78481 storage/store.go:1651  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181002 08:09:58.438611 78482 storage/store.go:1651  [s2,r1/2:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181002 08:09:58.467832 78388 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
I181002 08:09:58.481537 78724 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:33193
I181002 08:09:58.484212 78724 gossip/client.go:134  [n2] closing client to node 1 (127.0.0.1:33193): stopping outgoing client to node 1 (127.0.0.1:33193); already have incoming
W181002 08:09:58.704430 78386 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181002 08:09:59.165692 78387 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181002 08:09:59.412183 78388 storage/store.go:1651  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
    --- FAIL: TestRangeTransferLeaseExpirationBased/DrainTransferWithExtension (75.61s)
    	soon.go:49: condition failed to evaluate within 45s: expected replica with id 2 to be raft leader, instead got id 1
    		goroutine 78265 [running]:
    		runtime/debug.Stack(0xa7a358200, 0xc4213ca780, 0x3be1420)
    			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
    		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3c3d300, 0xc42020b590, 0xc420cad8c0)
    			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
    		github.com/cockroachdb/cockroach/pkg/storage_test.(*leaseTransferTest).ensureLeaderAndRaftState(0xc421d70100, 0xc42020b590, 0xc42048f180, 0x100000001, 0x1)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_replica_test.go:656 +0x1a7
    		github.com/cockroachdb/cockroach/pkg/storage_test.TestRangeTransferLeaseExpirationBased.func4(0xc42020b590)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_replica_test.go:855 +0x2a3
    		testing.tRunner(0xc42020b590, 0x34d01b0)
    			/usr/local/go/src/testing/testing.go:777 +0x16e
    		created by testing.(*T).Run
    			/usr/local/go/src/testing/testing.go:824 +0x565
    	client_test.go:361: timed out during shutdown

```

SHA: https://github.com/cockroachdb/cockroach/commits/842cd970e0e172d4604017f659c0a7b424dc4db7

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestRangeTransferLeaseExpirationBased PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1004907&tab=buildLog

```
W181109 09:56:20.393757 80581 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181109 09:56:20.415836 80689 storage/store.go:1652  [s2,r1/2:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181109 09:56:20.483192 80688 storage/store.go:1652  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181109 09:56:20.664576 80580 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181109 09:56:20.695985 80582 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
I181109 09:56:21.013371 80891 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:35021
I181109 09:56:21.018074 80891 gossip/client.go:134  [n2] closing client to n1 (127.0.0.1:35021): stopping outgoing client to n1 (127.0.0.1:35021); already have incoming
W181109 09:56:21.394807 80689 storage/store.go:1652  [s2,r1/2:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181109 09:56:21.529762 80581 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181109 09:56:21.555586 80580 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181109 09:56:21.611137 80688 storage/store.go:1652  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181109 09:56:21.681921 80582 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
I181109 09:56:22.013900 80902 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:35021
I181109 09:56:22.016383 80902 gossip/client.go:134  [n2] closing client to n1 (127.0.0.1:35021): stopping outgoing client to n1 (127.0.0.1:35021); already have incoming
W181109 09:56:22.327633 80689 storage/store.go:1652  [s2,r1/2:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181109 09:56:22.511198 80580 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181109 09:56:22.520463 80581 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181109 09:56:22.525206 80688 storage/store.go:1652  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181109 09:56:22.673914 80582 storage/store.go:1652  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
I181109 09:56:23.015004 80753 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:35021
I181109 09:56:23.018996 80753 gossip/client.go:134  [n2] closing client to n1 (127.0.0.1:35021): stopping outgoing client to n1 (127.0.0.1:35021); already have incoming
    --- FAIL: TestRangeTransferLeaseExpirationBased/DrainTransferWithExtension (75.53s)
    	soon.go:49: condition failed to evaluate within 45s: expected replica with id 2 to be raft leader, instead got id 1
    		goroutine 80470 [running]:
    		runtime/debug.Stack(0xa7a358200, 0xc421883b00, 0x3bd8f60)
    			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
    		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3c35ba0, 0xc4205932c0, 0xc4203823a0)
    			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
    		github.com/cockroachdb/cockroach/pkg/storage_test.(*leaseTransferTest).ensureLeaderAndRaftState(0xc420b28000, 0xc4205932c0, 0xc42085ee00, 0x100000001, 0x1)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_replica_test.go:656 +0x1a7
    		github.com/cockroachdb/cockroach/pkg/storage_test.TestRangeTransferLeaseExpirationBased.func4(0xc4205932c0)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_replica_test.go:855 +0x2a3
    		testing.tRunner(0xc4205932c0, 0x34cf028)
    			/usr/local/go/src/testing/testing.go:777 +0x16e
    		created by testing.(*T).Run
    			/usr/local/go/src/testing/testing.go:824 +0x565
    	client_test.go:361: timed out during shutdown

```

SHA: https://github.com/cockroachdb/cockroach/commits/1d5c68bced0c02f96de828d99cd4120ed67d359f

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stress TESTS=TestRangeTransferLeaseExpirationBased PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1045608&tab=buildLog

```
W181207 07:46:52.993501 81480 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181207 07:46:53.114904 81481 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181207 07:46:53.250985 81479 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181207 07:46:53.584775 81609 storage/store.go:1653  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
I181207 07:46:53.776668 81845 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:45477
I181207 07:46:53.781904 81845 gossip/client.go:134  [n2] closing client to n1 (127.0.0.1:45477): stopping outgoing client to n1 (127.0.0.1:45477); already have incoming
W181207 07:46:53.986764 81610 storage/store.go:1653  [s2,r1/2:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181207 07:46:54.026002 81481 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181207 07:46:54.079839 81480 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181207 07:46:54.243050 81479 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181207 07:46:54.680706 81609 storage/store.go:1653  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
I181207 07:46:54.777924 81744 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:45477
I181207 07:46:54.782833 81744 gossip/client.go:134  [n2] closing client to n1 (127.0.0.1:45477): stopping outgoing client to n1 (127.0.0.1:45477); already have incoming
W181207 07:46:54.893749 81610 storage/store.go:1653  [s2,r1/2:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181207 07:46:55.009024 81480 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip system config: node unavailable; try another peer
W181207 07:46:55.143230 81481 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
W181207 07:46:55.292536 81479 storage/store.go:1653  [s1,r1/1:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181207 07:46:55.680093 81609 storage/store.go:1653  [s2,r1/2:/M{in-ax}] could not gossip system config: node unavailable; try another peer
I181207 07:46:55.778838 81835 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:45477
I181207 07:46:55.784354 81835 gossip/client.go:134  [n2] closing client to n1 (127.0.0.1:45477): stopping outgoing client to n1 (127.0.0.1:45477); already have incoming
W181207 07:46:55.815674 81610 storage/store.go:1653  [s2,r1/2:/M{in-ax}] could not gossip node liveness: node unavailable; try another peer
    --- FAIL: TestRangeTransferLeaseExpirationBased/DrainTransferWithExtension (75.57s)
    	soon.go:49: condition failed to evaluate within 45s: expected replica with id 2 to be raft leader, instead got id 1
    		goroutine 81292 [running]:
    		runtime/debug.Stack(0xa7a358200, 0xc421359dd0, 0x3bf8940)
    			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
    		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3c55560, 0xc4207801e0, 0xc421970940)
    			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
    		github.com/cockroachdb/cockroach/pkg/storage_test.(*leaseTransferTest).ensureLeaderAndRaftState(0xc421ab2180, 0xc4207801e0, 0xc421a0d180, 0x100000001, 0x1)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_replica_test.go:656 +0x1a7
    		github.com/cockroachdb/cockroach/pkg/storage_test.TestRangeTransferLeaseExpirationBased.func4(0xc4207801e0)
    			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_replica_test.go:855 +0x2a3
    		testing.tRunner(0xc4207801e0, 0x34e93a0)
    			/usr/local/go/src/testing/testing.go:777 +0x16e
    		created by testing.(*T).Run
    			/usr/local/go/src/testing/testing.go:824 +0x565
    	client_test.go:361: timed out during shutdown

```

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -4436,6 +4436,8 @@ func fatalOnRaftReadyErr(ctx context.Context, expl string, err error) {
 // tick the Raft group, returning true if the raft group exists and is
 // unquiesced; false otherwise.
 func (r *Replica) tick(livenessMap map[roachpb.NodeID]bool) (bool, error) {
+	ctx := r.AnnotateCtx(context.TODO())
+
 	r.unreachablesMu.Lock()
 	remotes := r.unreachablesMu.remotes
 	r.unreachablesMu.remotes = nil
@@ -4458,9 +4460,11 @@ func (r *Replica) tick(livenessMap map[roachpb.NodeID]bool) (bool, error) {
 	if r.mu.quiescent {
 		return false, nil
 	}
-	if r.maybeQuiesceLocked(livenessMap) {
+	if r.maybeQuiesceLocked(ctx, livenessMap) {
 		return false, nil
 	}
+
+	r.maybeTransferRaftLeadershipLocked(ctx)
 
 	r.mu.ticks++
 	r.mu.internalRaftGroup.Tick()
@@ -4538,8 +4542,10 @@ func (r *Replica) tick(livenessMap map[roachpb.NodeID]bool) (bool, error) {
 // would quiesce. The fallout from this situation are undesirable raft
 // elections which will cause throughput hiccups to the range, but not
 // correctness issues.
-func (r *Replica) maybeQuiesceLocked(livenessMap map[roachpb.NodeID]bool) bool {
-	ctx := r.AnnotateCtx(context.TODO())
+
+func (r *Replica) maybeQuiesceLocked(
+	ctx context.Context, livenessMap map[roachpb.NodeID]bool,
+) bool {
 	status, ok := shouldReplicaQuiesce(ctx, r, r.store.Clock().Now(), len(r.mu.proposals), livenessMap)
 	if !ok {
 		return false
@@ -4553,28 +4559,45 @@ type quiescer interface {
 	raftLastIndexLocked() (uint64, error)
 	hasRaftReadyRLocked() bool
 	ownsValidLeaseRLocked(ts hlc.Timestamp) bool
-	maybeTransferRaftLeader(ctx context.Context, status *raft.Status, ts hlc.Timestamp)
 }
 
 func (r *Replica) hasRaftReadyRLocked() bool {
 	return r.mu.internalRaftGroup.HasReady()
 }
 
-// TODO(tschottdorf): there's also maybeTransferRaftLeadership. Only one should exist.
-func (r *Replica) maybeTransferRaftLeader(
-	ctx context.Context, status *raft.Status, now hlc.Timestamp,
-) {
-	l := *r.mu.state.Lease
-	if !r.isLeaseValidRLocked(l, now) {
-		return
-	}
+func (r *Replica) maybeTransferRaftLeadership(ctx context.Context) {
+	r.mu.Lock()
+	r.maybeTransferRaftLeadershipLocked(ctx)
+	r.mu.Unlock()
+}
+
+// maybeTransferRaftLeadershipLocked attempts to transfer the leadership away
+// from this node to the leaseholder, if this node is the current raft leader
+// but not the leaseholder. We don't attempt to transfer leadership if the
+// leaseholder is behind on applying the log.
+//
+// We like it when leases and raft leadership are collocated because that
+// facilitates quick command application (requests generally need to make it to
+// both the lease holder and the raft leader before being applied by other
+// replicas).
+func (r *Replica) maybeTransferRaftLeadershipLocked(ctx context.Context) {
 	if r.store.TestingKnobs().DisableLeaderFollowsLeaseholder {
 		return
 	}
-	if pr, ok := status.Progress[uint64(l.Replica.ReplicaID)]; ok && pr.Match >= status.Commit {
-		log.VEventf(ctx, 1, "transferring raft leadership to replica ID %v", l.Replica.ReplicaID)
+	lease := *r.mu.state.Lease
+	if lease.OwnedBy(r.StoreID()) || !r.isLeaseValidRLocked(lease, r.Clock().Now()) {
+		return
+	}
+	raftStatus := r.raftStatusRLocked()
+	if raftStatus == nil || raftStatus.RaftState != raft.StateLeader {
+		return
+	}
+	lhReplicaID := uint64(lease.Replica.ReplicaID)
+	lhProgress, ok := raftStatus.Progress[lhReplicaID]
+	if (ok && lhProgress.Match >= raftStatus.Commit) || r.mu.draining {
+		log.VEventf(ctx, 1, "transferring raft leadership to replica ID %v", lhReplicaID)
 		r.store.metrics.RangeRaftLeaderTransfers.Inc(1)
-		r.mu.internalRaftGroup.TransferLeader(uint64(l.Replica.ReplicaID))
+		r.mu.internalRaftGroup.TransferLeader(lhReplicaID)
 	}
 }
 
@@ -4621,12 +4644,6 @@ func shouldReplicaQuiesce(
 		if log.V(4) {
 			log.Infof(ctx, "not quiescing: not leaseholder")
 		}
-		// Try to correct leader-not-leaseholder condition, if encountered,
-		// assuming the leaseholder is caught up to the commit index.
-		//
-		// TODO(peter): It is surprising that a method named shouldReplicaQuiesce
-		// might initiate transfer of the Raft leadership.
-		q.maybeTransferRaftLeader(ctx, status, now)
 		return nil, false
 	}
 	// We need all of Applied, Commit, LastIndex and Progress.Match indexes to be

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -26,7 +26,6 @@ import (
 	"github.com/kr/pretty"
 	opentracing "github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
-	"go.etcd.io/etcd/raft"
 	"golang.org/x/time/rate"
 
 	"github.com/cockroachdb/cockroach/pkg/keys"
@@ -309,17 +308,10 @@ func (r *Replica) leasePostApply(ctx context.Context, newLease roachpb.Lease) {
 		r.txnWaitQueue.Clear(true /* disable */)
 	}
 
-	if !iAmTheLeaseHolder && r.IsLeaseValid(newLease, r.store.Clock().Now()) &&
-		!r.store.TestingKnobs().DisableLeaderFollowsLeaseholder {
-		// If this replica is the raft leader but it is not the new lease holder,
-		// then try to transfer the raft leadership to match the lease. We like it
-		// when leases and raft leadership are collocated because that facilitates
-		// quick command application (requests generally need to make it to both the
-		// lease holder and the raft leader before being applied by other replicas).
-		// Note that this condition is also checked periodically when computing
-		// replica metrics.
-		r.maybeTransferRaftLeadership(ctx, newLease.Replica.ReplicaID)
-	}
+	// If we're the current raft leader, may want to transfer the leadership to
+	// the new leaseholder. Note that this condition is also checked periodically
+	// when ticking the replica.
+	r.maybeTransferRaftLeadership(ctx)
 
 	// Notify the store that a lease change occurred and it may need to
 	// gossip the updated store descriptor (with updated capacity).
@@ -460,40 +452,6 @@ func addSSTablePreApply(
 	}
 	log.Eventf(ctx, "ingested SSTable at index %d, term %d: %s", index, term, path)
 	return copied
-}
-
-// maybeTransferRaftLeadership attempts to transfer the leadership
-// away from this node to target, if this node is the current raft
-// leader. We don't attempt to transfer leadership if the transferee
-// is behind on applying the log.
-//
-// TODO(tschottdorf): there's also maybeTransferRaftLeader. Only one should exist.
-func (r *Replica) maybeTransferRaftLeadership(ctx context.Context, target roachpb.ReplicaID) {
-	err := r.withRaftGroup(true, func(raftGroup *raft.RawNode) (bool, error) {
-		// Only the raft leader can attempt a leadership transfer.
-		if status := raftGroup.Status(); status.RaftState == raft.StateLeader {
-			// Only attempt this if the target has all the log entries. Although
-			// TransferLeader is supposed to do the right thing if the target is not
-			// caught up, this check avoids periods of 0 QPS:
-			// https://github.com/cockroachdb/cockroach/issues/22573#issuecomment-366106118
-			if pr, ok := status.Progress[uint64(target)]; (ok && pr.Match == r.mu.lastIndex) || r.mu.draining {
-				log.VEventf(ctx, 1, "transferring raft leadership to replica ID %v", target)
-				r.store.metrics.RangeRaftLeaderTransfers.Inc(1)
-				raftGroup.TransferLeader(uint64(target))
-			}
-		}
-		return true, nil
-	})
-	if err != nil {
-		// An error here indicates that this Replica has been destroyed
-		// while lacking the necessary synchronization (or even worse, it
-		// fails spuriously - could be a storage error), and so we avoid
-		// sweeping that under the rug.
-		//
-		// TODO(tschottdorf): this error is not handled any more
-		// at this level.
-		log.Fatal(ctx, roachpb.NewReplicaCorruptionError(err))
-	}
 }
 
 func (r *Replica) handleReplicatedEvalResult(

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -9782,13 +9782,6 @@ func (q *testQuiescer) ownsValidLeaseRLocked(ts hlc.Timestamp) bool {
 	return q.ownsValidLease
 }
 
-func (q *testQuiescer) maybeTransferRaftLeader(
-	ctx context.Context, status *raft.Status, ts hlc.Timestamp,
-) {
-	// Nothing to do here. We test Raft leadership transfer in
-	// TestTransferRaftLeadership.
-}
-
 func TestShouldReplicaQuiesce(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -1175,7 +1175,7 @@ func (s *Store) SetDraining(drain bool) {
 
 					if needsRaftTransfer {
 						r.raftMu.Lock()
-						r.maybeTransferRaftLeadership(ctx, drainingLease.Replica.ReplicaID)
+						r.maybeTransferRaftLeadership(ctx)
 						r.raftMu.Unlock()
 					}
 				}); err != nil {
@@ -2366,6 +2366,21 @@ func splitPostApply(
 	rightRng.mu.Unlock()
 	r.mu.Unlock()
 	log.Event(ctx, "copied timestamp cache")
+
+	// We need to explicitly wake up the Raft group on the right-hand range or
+	// else the range could be underreplicated for an indefinite period of time.
+	//
+	// Specifically, suppose one of the replicas of the left-hand range never
+	// applies this split trigger, e.g., because it catches up via a snapshot that
+	// advances it past this split. That store won't create the right-hand replica
+	// until it receives a Raft message addressed to the right-hand range. But
+	// since new replicas start out quiesced, unless we explicitly awaken the
+	// Raft group, there might not be any Raft traffic for quite a while.
+	if err := rightRng.withRaftGroup(true, func(r *raft.RawNode) (unquiesceAndWakeLeader bool, _ error) {
+		return true, nil
+	}); err != nil {
+		log.Fatalf(ctx, "unable to create raft group for right-hand range in split: %s", err)
+	}
 
 	// Invoke the leasePostApply method to ensure we properly initialize
 	// the replica according to whether it holds the lease. This enables

